internal/config/config.go

```diff
@@ -65,32 +66,45 @@ func RegisterConfigFile() error {
 	return nil
 }
 
-func GetConfig() *Config {
+func ResolveConfig() (*Config, error) {
+	// Collect allowed directories, so that paths in the config can be validated.
+	configDir := viperInstance.GetString(ConfigDirectoriesKey)
+	allowedDirs := make([]string, 0)
+	for _, dir := range strings.Split(configDir, ":") {
+		if dir != "" && filepath.IsAbs(dir) {
+			allowedDirs = append(allowedDirs, dir)
+		} else {
+			slog.Warn("Invalid directory: ", "dir", dir)
+		}
+	}
+
+	// Collect all parsing errors before returning the error, so the user sees all issues with config
+	// in one error message.
+	var err error
+	collector, otelcolErr := resolveCollector(allowedDirs)
+	err = errors.Join(err, otelcolErr)
+	if err != nil {
+		return nil, fmt.Errorf("invalid configuration: %w", err)
+	}
+
 	config := &Config{
 		UUID:               viperInstance.GetString(UUIDKey),
 		Version:            viperInstance.GetString(VersionKey),
 		Path:               viperInstance.GetString(ConfigPathKey),
-		Log:                getLog(),
-		DataPlaneConfig:    getDataPlaneConfig(),
-		Client:             getClient(),
-		ConfigDir:          getConfigDir(),
-		AllowedDirectories: []string{},
-		Metrics:            getMetrics(),
-		Command:            getCommand(),
-		Common:             getCommon(),
-		Watchers:           getWatchers(),
+		Log:                resolveLog(),
```

Comment:
I see what's going on now. I think having the collector have it's own function resolveCollector is better for consistency. It also simplifies the logic in this function. Splitting the resolves into their own functions seems like a cleaner and more testable approach

Reply:
Yeah I switched from Getters to use `Resolve*` naming just to get rid of the dreaded `Get` keyword 😄 The reason why `resolveCollector()` in particular is separate from the others here is that it now returns an `error` because it is now validated, which I think should be standard practice going forward.

Configs should be validated already at this stage so that we can return a failure early and prompt the user to fix a misconfigured Agent.